### PR TITLE
Gradual movement (regex query string, goto_next and goto_previous without specifying start / end)

### PR DESCRIPTION
**Pull request by kiyoon**
Before merging #359 I wanted to show you this.

At the moment, my keymaps are quite polluted.

```lua
      goto_next_start = {
        ["]m"] = "@function.outer",
        ["]f"] = "@call.outer",
        ["]d"] = "@conditional.outer",
        ["]o"] = "@loop.outer",
        ["]s"] = "@statement.outer",
        ["]a"] = "@parameter.outer",
        ["]c"] = "@comment.outer",
        ["]b"] = "@block.outer",
        ["]l"] = { query = "@class.outer", desc = "next class start" },
        ["]]m"] = "@function.inner",
        ["]]f"] = "@call.inner",
        ["]]d"] = "@conditional.inner",
        ["]]o"] = "@loop.inner",
        ["]]a"] = "@parameter.inner",
        ["]]b"] = "@block.inner",
        ["]]l"] = { query = "@class.inner", desc = "next class start inner" },
      },
      goto_next_end = {
        ["]M"] = "@function.outer",
        ["]F"] = "@call.outer",
        ["]D"] = "@conditional.outer",
        ["]O"] = "@loop.outer",
        ["]S"] = "@statement.outer",
        ["]A"] = "@parameter.outer",
        ["]C"] = "@comment.outer",
        ["]B"] = "@block.outer",
        ["]L"] = "@class.outer",
        ["]]M"] = "@function.inner",
        ["]]F"] = "@call.inner",
        ["]]D"] = "@conditional.inner",
        ["]]O"] = "@loop.inner",
        ["]]A"] = "@parameter.inner",
        ["]]B"] = "@block.inner",
        ["]]L"] = "@class.inner",
      },
      goto_previous_start = {
        ["[m"] = "@function.outer",
        ["[f"] = "@call.outer",
        ["[d"] = "@conditional.outer",
        ["[o"] = "@loop.outer",
        ["[s"] = "@statement.outer",
        ["[a"] = "@parameter.outer",
        ["[c"] = "@comment.outer",
        ["[b"] = "@block.outer",
        ["[l"] = "@class.outer",
        ["[[m"] = "@function.inner",
        ["[[f"] = "@call.inner",
        ["[[d"] = "@conditional.inner",
        ["[[o"] = "@loop.inner",
        ["[[a"] = "@parameter.inner",
        ["[[b"] = "@block.inner",
        ["[[l"] = "@class.inner",
      },
      goto_previous_end = {
        ["[M"] = "@function.outer",
        ["[F"] = "@call.outer",
        ["[D"] = "@conditional.outer",
        ["[O"] = "@loop.outer",
        ["[S"] = "@statement.outer",
        ["[A"] = "@parameter.outer",
        ["[C"] = "@comment.outer",
        ["[B"] = "@block.outer",
        ["[L"] = "@class.outer",
        ["[[M"] = "@function.inner",
        ["[[F"] = "@call.inner",
        ["[[D"] = "@conditional.inner",
        ["[[O"] = "@loop.inner",
        ["[[A"] = "@parameter.inner",
        ["[[B"] = "@block.inner",
        ["[[L"] = "@class.inner",
      },
```

This patch enables you to group things together if you wish.  
```lua
      goto_next = {
        ["]m"] = "@function.*",
        ["]f"] = "@call.*",
        ["]o"] = { query = { "@loop.*", "@conditional.*" }, desc = "Next loop or conditional" },
        ["]s"] = "@statement.*",
        ["]a"] = "@parameter.*",
        ["]c"] = "@comment.*",
        ["]b"] = "@block.*",
        ["]l"] = { query = "@class.*", desc = "next class start/end" },
      },
      goto_previous = {
        ["[m"] = "@function.*",
        ["[f"] = "@call.*",
        ["[o"] = { query = { "@loop.*", "@conditional.*" }, desc = "Previous loop or conditional" },
        ["[s"] = "@statement.*",
        ["[a"] = "@parameter.*",
        ["[c"] = "@comment.*",
        ["[b"] = "@block.*",
        ["[l"] = { query = "@class.*", desc = "previous class start/end" },
      },
```

(I just noticed that you can already add multiple queries, however it isn't documented)  

Along with the power of repeatable movements, for some people it may be a good bonus as they can go to the right part they want without going for a complicated keymapping.

The regex queries are supported in the original `goto_next_start` etc. as well.  
It will stay mostly compatible with the previous setup, except that now the dot will match any character. If you have a custom query named `@functionQinner` then it will be matched with `@function.inner`

### CODE REVIEW

**Comment by theHamsta** [comment]
How complicated would it be it integrate https://github.com/nvim-treesitter/nvim-treesitter-textobjects/pull/60/files in your PR for `move`?

File: lua/nvim-treesitter/textobjects/move.lua [hunk 1]
```diff
@@ -33,39 +73,35 @@ local function move(query_strings, forward, start, winid)
...
+      for _, start_ in ipairs(starts) do
+        local current_match = queries.find_best_match(bufnr, query_string, "textobjects", function(match)
+          return filter_function(start_, match)
+        end, function(match)
+          return scoring_function(start_, match)
+        end)
+
+        if current_match then
+          local score = scoring_function(start_, current_match)
+          print(score)
```
Comment (theHamsta) [inline lua/nvim-treesitter/textobjects/move.lua:90]: debug statement

**Comment by kiyoon** [comment]
> How complicated would it be it integrate https://github.com/nvim-treesitter/nvim-treesitter-textobjects/pull/60/files in your PR for `move`? At the moment, you always call `"textobjects"` for `find_best_match`, would it be possible to allow queryies from other queries types as well?

Any reason it hasn't been merged?

Where should the `locals.scm` file be located in order to test this?

File: lua/nvim-treesitter/textobjects/move.lua (hunk 1 above)
Comment (kiyoon) [reply lua/nvim-treesitter/textobjects/move.lua]: Done

**Comment by theHamsta** [comment]
> > How complicated would it be it integrate https://github.com/nvim-treesitter/nvim-treesitter-textobjects/pull/60/files in your PR for `move`? At the moment, you always call `"textobjects"` for `find_best_match`, would it be possible to allow queryies from other queries types as well?
> 
> Any reason it hasn't been merged?
> 
> Where should the `locals.scm` file be located in order to test this?

It isn't about `"locals"` in specifc, in order for textobjects to work, you need to have `queries/<lang>/textobjects.scm` in you path. The idea of this PR is to also allow `queries/<lang>/<something>.scm`. `nvim-treesitter` provides queries for `highlights.scm`/`locals.scm`/`indent.scm`/`folds.scm`. So all of these should work. They just need to be in your runtimepath, e.g. relative to some plugin path.

I was a bit hesitant whether we should allow this right now and now it is a bit out-dated. But since you're touching the config anyways you could also make it more general. At least it should be possible to add alternative query groups in retroperspecitve

**Comment by kiyoon** [comment]
@theHamsta I see, the scope movement is actually amazing!  
Now I added the query group to the movement.  
It is only in movement atm and not in others yet.

**Comment by kiyoon** [comment]
Should I make multiple query groups selectable? I feel like this will make it super complicated and most people wouldn't really select from multiple groups?

**Comment by kiyoon** [comment]
Finished integrating #60

**Comment by kiyoon** [comment]
Tests passing

**Review by theHamsta — APPROVED** [review]
Wow, cool! I actually only meant to consider it for movement to see whether the config stays compatible without adding too much code. But cool that you made it working for all categories

File: lua/nvim-treesitter/textobjects/select.lua [hunk 2]
```diff
@@ -145,33 +146,52 @@ function M.detect_selection_mode(query_string, keymap_mode)
...
+    if not available then
+      query_string = nil
+    end
+
+    if query_string then
+      local cmd_o = function()
+        M.select_textobject(query_string, query_group, "o")
+      end
+      local cmd_x = function()
+        M.select_textobject(query_string, query_group, "x")
```
Comment (theHamsta) [inline lua/nvim-treesitter/textobjects/select.lua:185]: @kiyoon weren't lua functions causing problems in visual mode? There was a reason why this still was a string.

Comment (kiyoon) [reply lua/nvim-treesitter/textobjects/select.lua:185]: @theHamsta I don't know, I think maybe before the patch from nvim-treesitter there were lots of weird bugs even with `<cmd>` and now it seems okay. It at least passes the tests with visual selection

Comment (theHamsta) [reply lua/nvim-treesitter/textobjects/select.lua:185]: that's good. glad that we can finally get rid of this vimscript artifact.

Comment (kiyoon) [reply lua/nvim-treesitter/textobjects/select.lua:185]: @theHamsta By the way, when does the `detach` function exactly called? I think my workflow is usually limited to having not many buffers at the same time, but it doesn't seem like it's being called whenever you close the buffer? Or I am doing something wrong

**Pull request merged**